solidity/nft_token.sol

```diff
@@ -0,0 +1,254 @@
+            revert("Wrong NFT address");
+        }
+
+        // Check tick values
+        if (position.tick_lower_index != min_tick_lower_index || position.tick_upper_index != max_tick_lower_index) {
+            revert("Wrong ticks");
+        }
+
+        // Check the PDA ownership
+        // TODO: Uncomment following lines in production
+//        if (ai.owner != orca) {
+//            revert("Wrong pda owner");
+//        }
+
+        // Check the PDA header data
+        uint64 header = ai.data.readUint64LE(0);
+        if (header != pdaHeader) {
+            revert("Wrong pda header");
+        }
+
+        // Check the PDA address correctness
+        (address pdaPosition, ) = try_find_program_address(["position", position.position_mint], orca);
+        // TODO: Uncomment following lines in production
+//        if (pdaPosition != positionDataAccount) {
+//            revert("Wrong position pda");
+//        }
+
+        // TODO: Do the liquidity check for max(uint64) value as it is provided as uint128 from the LP provider
+        uint64 liquidity = uint64(position.liquidity);
+
+        // Transfer the NFT to the pdaTokenAccount address of this program
+        SplToken.transfer(
+            tx.accounts.fromTokenAccount.key,
+            tx.accounts.pdaTokenAccount.key,
+            tx.accounts.fromWallet.key,
+            1);
+
+        // Transfer ERC20 tokens to the user
+        SplToken.mint_to_pda(
+            mintErc20,
+            tx.accounts.toErc20.key,
+            pdaProgram,
+            liquidity,
+            pdaProgramSeed,
+            pdaBump);
+
+        // Record liquidity and its correspondent token account address
+        address tokenAccount = tx.accounts.pdaTokenAccount.key;
+        mapTokenAccountLiquidity[tokenAccount] = liquidity;
+        tokenAccounts[numTokenAccounts] = (tx.accounts.pdaTokenAccount.key);
+        numTokenAccounts++;
+    }
+
+    @mutableAccount(fromERC20Account)
+    @mutableAccount(pdaERC20Account)
+    @mutableAccount(fromWallet)
+    @mutableAccount(pdaTokenAccount)
+    @mutableAccount(fromTokenAccount)
+    @mutableAccount(mintERC20)
+    @signer(sig)
+    // Transfer with PDA
+    function withdraw(uint64 amount) external {
+        // Check that the token account exists
+        if (mapTokenAccountLiquidity[tx.accounts.pdaTokenAccount.key] == 0) {
+            revert("No liquidity on a provided token account");
+        }
+
+        // Transfer ERC20 tokens to the pdaERC20Account address of this program
+        SplToken.transfer(
+            tx.accounts.fromERC20Account.key,
+            tx.accounts.pdaERC20Account.key,
+            tx.accounts.fromWallet.key,
+            amount);
+
+        // Transfer NFT to the user associated token account
+        SplToken.transfer_pda(
+            tx.accounts.pdaTokenAccount.key,
+            tx.accounts.fromTokenAccount.key,
+            pdaProgram,
+            1,
+            pdaProgramSeed,
+            pdaBump);
```

Comment:
Don't we have to account for the potential splitting of the liquidity here?

Reply:
This part is not yet complete, but we are certain now that we don't need to do that, as decreaseLiquidity() function removes the specified liquidity from a specific position. This way we don't need to completely liquidate a full position and create another one from remainders later.